design/new-prepost-backuprestore-plugin-hooks.md

```diff
@@ -0,0 +1,713 @@
+
+Other examples of plugins that can use the proposed plugin hooks are:
+
+- PostBackupAction: trigger a Velero Restore after a successful Velero backup (and complete the migration operation).
+- PreRestoreAction: pre-expand the cluster's capacity via Cluster API to avoid starvation of cluster resources before the restore.
+- PostRestoreAction: call actions to be performed outside Kubernetes clusters, such as configure a global load balancer (GLB) that enables the new cluster.
+
+The post backup actions will be executed after the backup is uploaded (persisted) on the disk. The logs of post-backup actions will be uploaded on the disk once the actions are completed.
+
+The post restore actions will be executed after the restore is uploaded (persisted) on the disk. The logs of post-restore actions will be uploaded on the disk once the actions are completed.
+
+This design seeks to provide missing extension points. This proposal's scope is to only add the new plugin hooks, not the plugins themselves.
+
+## Goals
+
+- Provide PreBackupAction, PostBackupAction, PreRestoreAction, and PostRestoreAction APIs for plugins to implement.
+- Update Velero backup and restore creation logic to invoke registered PreBackupAction and PreRestoreAction plugins before processing the backup and restore respectively.
+- Update Velero backup and restore complete logic to invoke registered PostBackupAction and PostRestoreAction plugins the objects are uploaded on disk.
+- Create one `ActionStatus` struct to keep track of execution of the plugin hooks. This struct has PluginName, StartTimestamp, CompletionTimestamp and Phase.
+- Add sub statuses for the plugins on Backup object: `Backup.Status.PreBackupActionsStatuses` and `Backup.Status.PostBackupActionsStatuses`. They will be flagged as optional and nullable. They will be populated only each plugin registered for the PreBackup and PostBackup hooks, respectively.
+- Add sub statuses for the plugins on Restore object: `Backup.Status.PreRestoreActionsStatuses` and `Backup.Status.PostRestoreActionsStatuses`. They will be flagged as optional and nullable. They will be populated only each plugin registered for the PreRestore and PostRestore hooks, respectively.
+- that will be populated optionally if Pre/Post Backup/Restore.
+
+## Non-Goals
+
+- Specific implementations of the PreBackupAction, PostBackupAction, PreRestoreAction and PostRestoreAction API beyond test cases.
+- For migration specific actions (Velero Backup + Velero Restore), add disk synchronization during the validation of the Restore (making sure the newly created backup will show during restore)
+
+## High-Level Design
```

Comment:
A couple of questions:
1.  If a PostBackup action fails, do we consider the backup successful and usable?
2. How does user get the execution log of PostBackup plugin?
3. Once registered, will the plugins per backup/restore take effect to all backups and restores?  Will there be a mechanism such these plugins can be executed selectively?
4. How does user define timeout?  And how will it be handled?

Reply:
Answering:
1. Yes
2. The plan is to implement an interface to persist the post backup log on the same location as the backup log (not necessarily inside the gzip file, but on the same directory as a different name).
3. Yes. But the plugin developer can implement on their plugin a mechanism so the plugin to execute selectively, for example, a config map. 
4. This design does not touch on timeouts since current hooks do not provide any provisioning for such. There is an en-route design about adding timeout on plugins, which I propose to bring back to the discussion table once we have these new hooks implemented: https://github.com/vmware-tanzu/velero/pull/3355 .

Reply:
> The plan is to implement an interface to persist the post backup log on the same location as the backup log (not necessarily inside the gzip file, but on the same directory as a different name).

If we provide a way for users to collect the backup log, we should also provide a way for users to see the log of the plugin executions.

For 3, if we agree that for most users the plugins should not always be triggered for different backup and restores we should provide a way in the framework to enable user to choose what plugins to be executed or skipped.  IMO, Iinstead of relying on users to workaround, it should be covered in the design.

Reply:
On point #4, when we create the post backup log, we will make sure the `velero` command line will be able to collect them. For point #3, do we have similar functionality on existent hooks + plugins? If not, I am glad to put some design doc after this one goes through.

Reply:
@brito-rafa 

> On point #4, when we create the post backup log, we will make sure the velero command line will be able to collect them. 

I think you mean point 2, it's probably OK to put it in another PR to make velero CLI return _additional logs_ as it's also needed for itemSnapshotter plugins.

But for point 3, the existing hooks are at a resource level, which doesn't impact all backups and restores, additionally, the user who creates the backup can edit the annotations to control whether the hooks will be executed.  We need to address that for the pre/post backup/restore plugin before this feature is delivered.  Therefore, ideally, it should be covered in this design.

Reply:
I concur on annotations. I added a section on the design document about this.